src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts

```diff
@@ -132,4 +136,59 @@ export class TableStructureComponent implements OnChanges {
     return unique;
   }
 
+  classifyComplexesSimilaritiesV2(complexesList: Element[]) {
+    const comparedComplexes: [Element, Element, number][] = [];
+    const finalList: Element[] = [];
+
+    for (const complex of complexesList) {
+      for (const comparedComplex of complexesList) {
+        // for unique comparison
+        if (complex.complexAC >= comparedComplex.complexAC) {
+          comparedComplexes.push([complex, comparedComplex, this.calculateSimilarity(complex, comparedComplex)]);
+        }
+      }
+    }
+    comparedComplexes.sort((a, b) => b[2] - a[2]); // sorting by similarityScore
+
+    // const complexesOrderedSet = new Set<Element>(); // used to have unique complexes
+    // for (const [complex1, complex2] of comparedComplexes) {
+    //   if (!complexesOrderedSet.has(complex1)) {
+    //     complexesOrderedSet.add(complex1);
+    //   }
+    //   if (!complexesOrderedSet.has(complex2)) {
+    //     complexesOrderedSet.add(complex2);
+    //   }
+    // }
+
+    const complexesOrderedSet = this.uniqueComplexesListOrderedBySimilarity(comparedComplexes);
+    // to be used in the table as a 1D array
+    for (const complex of complexesOrderedSet) {
+      finalList.push(complex);
+    }
+    return finalList;
+  }
+
+  uniqueComplexesListOrderedBySimilarity(complexesListSimilarities: [Element, Element, number][]) {
+    const complexesOrderedSet = new Set<Element>();
```

Comment:
comparedComplexes is sorting the comparedComplexes Array by the highest similarity score before creating the complexesOrderedSet using uniqueComplexesListOrderedBySimilarity,  isn't it enough?

Reply:
We sort `complexesListSimilarities` before this method, so when we do `for (const [complex1, complex2, similarityScore] of complexesListSimilarities)` we are iterating through them in order.

Based on that, we insert into `comparedComplexes`, also in order. But `comparedComplexes` is a Set, and a set guarantees objects are unique and quick to access, but it does not guarantee it's in order of insertion.

So, when we do `for (const complex of complexesOrderedSet) {` (line 165), we are not guaranteed to iterate through the complexes in order.

We have 2 options:

1. Find an implementation of set that it is ordered. If we find one, we also don't need checks like the following, as set already doesn't allow duplicates.
    ```
    if (!complexesOrderedSet.has(complex1)) {
      complexesOrderedSet.add(complex1);
    }
    ```

2. Use a list. In line 172, we can do `const complexesOrderedSet: Element[] = [];` and it would work without any other changes. If we do this, we also don't need `finalList`, as we already have a list.

Reply:
I had a look at this, and it looks like Set returns the items ordered by insertion, in which case, using a set is fine. Just a few changes.

1. We don't need to call `has`, so instead of
   ```
   if (!complexesOrderedSet.has(complex1)) {
      complexesOrderedSet.add(complex1);
   }
   ```
   You can just do
   ```
   complexesOrderedSet.add(complex1);
   ```

2. Simply making the set into a list. Instead of
   ```
    for (const complex of complexesOrderedSet) {
      finalList.push(complex);
    }
    return finalList;
   ```
   You can do
   ```
   return Array.from(complexesOrderedSet);
   ```

3. Remove the commented out code.